Add ChineseCalendar.prototype.monthsInYear

n-riesco opened the pull request:
This fixes #6 but it doesn't address the issue with `cc.add(dc, 1, 'y')` (I'll have a look at it now).

n-riesco committed:
chinese: override method monthsInYear

Fixes #6

n-riesco committed:
test: chineseCalendar.monthsInYear

n-riesco committed:
chinese: override method `add`

* Ensure month is kept in sync when adding years.

n-riesco committed:
test: chinese month is sync'ed after adding years

n-riesco commented:
This PR closes #6 completely:
- fixes the creation of dates in the 12th month of a leap year
- ensures that adding years keeps the initial Chinese month

alexcjohnson commented on jquery-src/jquery.calendars.chinese.js:
```diff
@@ -313,12 +313,21 @@
             firstThursday.add(4 - (firstThursday.dayOfWeek() || 7), 'd');
 
-            // compute days from first Thrusday
+            // compute days from first Thursday
```
haha I noticed this but didn't want to be too big a pain 🤓 :

alexcjohnson commented on jquery-src/jquery.calendars.chinese.js:
```diff
@@ -273,7 +273,7 @@
             var intercalaryMonth = monthDaysTable >> 13;
 
-            return (intercalaryMonth && intercalaryMonth === monthIndex);
+            return (intercalaryMonth !== 0 && intercalaryMonth === monthIndex);
```
bugfix or style?

n-riesco replied:
Strictly speaking is a bug fix because `0 || something` evaluates to `0` instead of `false`, and I've documented this function to return a boolean.

n-riesco replied:
I'm not happy with this function returning a `boolean`. Originally, I decided to return a `boolean` to resemble `leapYear`, but I often find myself wanting to know the intercalary month.

How do you think? After the dust settles, shall I change it to return the intercalary month number and 0 if none. This change is mostly backwards- compatible, and it'd help me avoid unnecessary recalculations in some of the methods.

alexcjohnson replied:
"mostly"? If 0 is not a possible intercalary month (which looks to me like a true statement), then returning something falsy (might as well be zero) or the month if there is one seems more useful, yes.

n-riesco replied:
@alexcjohnson Shall make the change in this PR?

alexcjohnson replied:
sure, might as well.

n-riesco replied:
@alexcjohnson Actually, the change is a bit more intrusive.

We would go from:

```js
var isIntercalary = chineseCalendar.intercalaryMonth(year, monthIndex);
```

to:

```js
var isIntercalary = (monthIndex === chineseCalendar.intercalaryMonth(year));
```

I still think the change is for better.

alexcjohnson commented on test/test.js:
```diff
@@ -236,4 +236,39 @@ describe('Chinese calendar', function() {
+            expect(resultYear).toEqual(year + 1);
+            expect(resultMonth).toEqual(month);
+            expect(resultIsIntercalary)
+                .toEqual(isIntercalary && resultLeapYear);
```
you're not hitting this case, are you? because you're only adding one year in the test, so you'll never have both the start and end year be leap years. But if you did I don't think this is the right test - you should only increment from intercalary to intercalary if it's the *same* intercalary month in the end year.

n-riesco replied:
No, I'm not, because I'm adding only 1 year.

n-riesco replied:
> you should only increment from intercalary to intercalary if it's the same intercalary month in the end year

Good catch, I'll fix that!

n-riesco committed:
chinese: fix `add` from leap month to leap year

* Ensure that the result of adding years to a date is an intercalary
  month, only if the start month is intercalary and the end year has the
  same intercalary month.

n-riesco committed:
chinese: change signature of intercalaryMonth

* ChineseCalendar.prototype.intercalaryMonth(year) returns the number of
  the intercalary month in a year or 0 if none.
